pkg/cmd/roachtest/tests/mixed_version_cdc.go

```diff
@@ -343,12 +348,60 @@ func (cmvt *cdcMixedVersionTester) createChangeFeed(node int) versionStep {
 			{"updated", ""},
 			{"resolved", fmt.Sprintf("'%s'", resolvedInterval)},
 		}
-		_, err := newChangefeedCreator(db, fmt.Sprintf("%s.%s", targetDB, targetTable), cmvt.kafka.sinkURL(ctx)).
-			With(options...).
-			Create()
+		creator := newChangefeedCreator(db, fmt.Sprintf("%s.%s", targetDB, targetTable), cmvt.kafka.sinkURL(ctx)).
+			With(options...)
+
+		jobID, err := creator.Create()
 		if err != nil {
 			t.Fatal(err)
 		}
+		// In 22.1, a changefeed may fail during a rolling upgrade due to a bug
+		// which is fixed in 22.2 onwards. To accomodate for this bug,
+		// poll the changefeed job and restart it if it fails.
+		// See #106878 for more details.
+		if cmvt.terminateWatcherCh == nil {
+			cmvt.terminateWatcherCh = make(chan struct{})
+		}
+		cmvt.monitor.Go(func(ctx2 context.Context) error {
+			watchChangefeed(ctx2, t, cmvt.terminateWatcherCh, jobID, db, creator)
+			return nil
+		})
+	}
+}
+
+// watchChangefeed polls a changefeed job every 1 second and re-creates if it
+// failed.
+func watchChangefeed(
+	ctx context.Context,
+	t test.Test,
+	testFinished chan struct{},
+	jobID int,
+	db *gosql.DB,
+	creator *changefeedCreator,
+) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-testFinished:
+			return
+		case <-time.After(1 * time.Second):
+			var feedStatus string
+			if err := db.QueryRow(`SELECT status FROM system.jobs WHERE id = $1`, jobID).Scan(&feedStatus); err != nil {
+				t.Fatal(errors.Wrap(err, "changefeed watcher failed to query jobs table"))
```

Comment:
Won't this flake if the query runs concurrently with node restarts? We can probably make this loop be best-effort: log any errors found, and try again on the next iteration.